core/server/worker/src/main/java/alluxio/worker/block/SpecificMasterBlockSync.java

```diff
@@ -0,0 +1,248 @@
+   * @param workerAddress the net address of the worker
+   * @param masterClientPool the Alluxio master client pool
+   * @param heartbeatReporter the heartbeat reporter
+   */
+  public SpecificMasterBlockSync(
+      BlockWorker blockWorker, AtomicReference<Long> workerId,
+      WorkerNetAddress workerAddress, BlockMasterClientPool masterClientPool,
+      BlockHeartbeatReporter heartbeatReporter)
+      throws IOException {
+    mBlockWorker = blockWorker;
+    mWorkerId = workerId;
+    mWorkerAddress = workerAddress;
+    mMasterClientPool = masterClientPool;
+    mMasterClient = mMasterClientPool.acquire();
+    mAsyncBlockRemover = new AsyncBlockRemover(mBlockWorker);
+    mBlockMasterSyncHelper = new BlockMasterSyncHelper(mMasterClient);
+    mMasterAddress = masterClientPool.acquire().getRemoteSockAddress();
+    mBlockHeartbeatReporter = heartbeatReporter;
+  }
+
+  private void registerWithMaster() {
+    RetryPolicy retry = createEndlessRetry();
+    while (retry.attempt()) {
+      try {
+        LOG.info("Registering with master {}", mMasterAddress);
+        registerWithMasterInternal();
+        LOG.info("Finished registration with {}", mMasterAddress);
+        return;
+      } catch (Exception e) {
+        LOG.error("Failed to register with master {}, error {}, retry count {} Will retry...",
+            mMasterAddress, e, retry.getAttemptCount());
+        mWorkerState = WorkerMasterState.NOT_REGISTERED;
+      }
+    }
+    // Should not reach here because the retry is indefinite
+    ProcessUtils.fatalError(LOG, new RuntimeException(),
+        "Failed to register with master %s", mMasterAddress);
+  }
+
+  private void registerWithMasterInternal()
+      throws IOException, FailedToAcquireRegisterLeaseException {
+    // The target master is not necessarily the one that allocated the workerID
+    LOG.info("Notify the master {} about the workerID {}", mMasterAddress, mWorkerId);
+    mMasterClientPool.acquire().addWorkerId(mWorkerId.get(), mWorkerAddress);
+    BlockStoreMeta storeMeta = mBlockWorker.getStoreMetaFull();
+
+    try {
+      mBlockMasterSyncHelper.tryAcquireLease(mWorkerId.get(), storeMeta);
+    } catch (FailedToAcquireRegisterLeaseException e) {
+      if (Configuration.getBoolean(PropertyKey.TEST_MODE)) {
+        throw new RuntimeException(String.format("Master register lease timeout exceeded: %dms",
+            ACQUIRE_LEASE_WAIT_MAX_DURATION));
+      }
+      throw e;
+    }
+    mWorkerState = WorkerMasterState.REGISTERING;
+    mBlockMasterSyncHelper.registerToMaster(mWorkerId.get(), storeMeta);
+    mWorkerState = WorkerMasterState.REGISTERED;
+    mLastSuccessfulHeartbeatMs = CommonUtils.getCurrentMs();
+  }
+
+  private RetryPolicy createEndlessRetry() {
+    return new ExponentialBackoffRetry(
+        1000, 60 * 1000, Integer.MAX_VALUE);
+  }
+
+  @Override
+  public synchronized void heartbeat() throws InterruptedException {
```

Comment:
endless retries are used in worker registration and heartbeats as we don't require heartbeats to all masters to succeed